### ISSUE TITLE
prov/verbs misleading use of ibv_fork_init()

### ISSUE THREAD

**Issue by soumagne** [ofiwg/libfabric#4974]
As initially reported in https://github.com/mercury-hpc/mercury/issues/280 we recently discovered some weird interactions when using both MPI (MVAPICH) and libfabric with verbs;ofi_rxm on one of our systems.
After some testing, we discovered that by default the verbs provider calls `ibv_fork_init()` unless the environment variable `FI_FORK_UNSAFE`  has been set to 1.
The problem is though that in the case of an interaction with MPI, the MPI launcher/library may have already made some ibv calls, which makes that `ibv_fork_init()` call to fail as it needs to be called first.
Not only that environment variable seems to be duplicating already existing environment variables that are `IBV_FORK_SAFE` or `RDMAV_FORK_SAFE`, but also the default behavior should not be to provide fork safety since `ibv_fork_init()` seems to also have performance implications.

In summary :) it would seem better both from an interaction and performance point of view to have the verbs provider not calling `ibv_fork_init()` at all and not providing its own `FI_FORK_UNSAFE` env variable but instead let users make use of `IBV_FORK_SAFE` or `RDMAV_FORK_SAFE` variables when needed, since it can be too late anyway for the verbs provider to issue that call.

**Comment by shefty** [comment]
@a-ilango @jswaro - fyi - we were just talking about this

**Comment by a-ilango** [comment]
https://github.com/ofiwg/libfabric/issues/1763 - this seems to be original issue that required enabling fork safety by default. I think the reason was to support more apps by default. @shefty do we still want to honor that? I think we should revert the patch if it's causing performance / functional issues.

**Comment by jswaro** [comment]
@shefty Thanks for the heads up. 
@soumagne see #4969 -- related, but not directly. 

I guess this doesn't surprise me. I recall a version of MPI or two that used some IBV calls to probe Infiniband devices prior to opening the communicator -- either for determine what the device is capable of, and/or determining what comm layer abstractions are available. 

This comes back to the discussion in #4969 about application intent, with a bit of mixed-model problems thrown in.  

A solution, even though it crosses the boundary that libfabric is responsible for, would seem to be removing the FI_FORK_UNSAFE variable. However, this comes at a cost for applications. Applications would need to know what environment variables need to be set for a given provider's underlying libraries rather than allowing the provider to handle it -- and that's because of the mixed model aspect where MPI is querying Infiniband directly rather than solely using libfabric.

**Comment by shefty** [comment]
There seem to be other issues related to how fork safe the library is anyway.  We may not have a decent alternative here.

**Comment by soumagne** [comment]
I think requesting fork safety is very specific and it seems easier for a user to look for a `FI_FORK_SAFE` variable at the very least, knowing that it is needed and assuming that it may fail if the parent process has called into ibv before. It is definitely not intuitive to look for the opposite `FI_FORK_UNSAFE` variable where users get safety for something specific that they have not really asked for or will most likely not need.

**Comment by shefty** [comment]
I don't believe the original intent was for libfabric to support being usable across a fork() call.  There's an issue that an app may make a system call which could invoke fork() under the covers.  So, the app developer may not even know if fork() support is required.  I've never seen performance data that shows the difference is between the optimal/no-fork option versus safe/fork option.

Without having performance data, I don't see a compelling reason to revisit the design decision on disabling fork() support by default.  If ibv_fork_init() fails, the failure should be logged, but I don't believe this should result in disabling the verbs provider.  Most apps don't call fork(), and we should be able to proceed.

@a-ilango - I think we can safely drop the 'goto done' in fi_ibv_init_info() on the ibv_fork_init() failure case.

The existing FI_FORK_UNSAFE variable should be usable as a work-around until then.

**Comment by shefty** [comment]
See #4973 and #4969 for related details.

Performance data is there.  Based on that data, we should disable fork support by default for performance gains.

**Comment by shefty** [comment]
Direct calls to ibv_fork_init() have been removed.  The only viable option is to use the verbs environment variable directly.  This is captured in the fi_verbs.7 man page.

**Comment by soumagne** [comment]
Great, thanks!
